### PR TITLE
[Meta] list for metadata

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Replace hashtable (glib) to list, setting/getting metadata of edge handle.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

File: src/libnnstreamer-edge/nnstreamer-edge-common.c [hunk 1]
```diff
@@ -138,6 +138,276 @@ nns_edge_strdup_printf (const char *format, ...)
...
+    return NNS_EDGE_ERROR_INVALID_PARAMETER;
+
+  nns_edge_metadata_free (dest);
+
+  node = src->list;
+  while (node) {
+    ret = nns_edge_metadata_set (dest, node->key, node->value);
+    if (ret != NNS_EDGE_ERROR_NONE) {
+      nns_edge_metadata_free (dest);
+      return ret;
```
Comment (myungjoo) [inline src/libnnstreamer-edge/nnstreamer-edge-common.c:301]: nns_edge_metadata_s tmp
tmp.list = dest->list;
loop: set(&tmp, ...)
if fail: return w/ error code (not touching dest)
if success: free dest, set dest->list = tmp.list

File: src/libnnstreamer-edge/nnstreamer-edge-common.h [hunk 2]
```diff
@@ -45,6 +45,28 @@ extern "C" {
...
+  nns_edge_metadata_node_s *next;
+};
+
+/**
+ * @brief Internal data structure to handle metadata. This struct should be managed in the handle.
+ */
+typedef struct {
+  unsigned int list_len;
+  nns_edge_metadata_node_s *list;
+} nns_edge_metadata_s;
```
Comment (myungjoo) [inline src/libnnstreamer-edge/nnstreamer-edge-common.h:68]: Need a lock to protect if the functions of nnstreamer-edge-common can be accessed from different threads.

Comment (again4you) [reply src/libnnstreamer-edge/nnstreamer-edge-common.h:68]: I agree with myungjoo's opinion.  The result of `nns_edge_metadata_find()` would vary if `nns_edge_metadata_set()` function is running in another thread. I thinks that the read lock mechanism is good enough for this issue.

**Review by again4you — COMMENTED** [review]
The code is good to read and easy to understand. All primitive functions based on Glib are changed to non-Glib functions so we can port `nnstreamer-edge` to other platforms more easily.

I just add some comments about multi-threading, memory function, and early exit. Please check my opinion.

File: src/libnnstreamer-edge/nnstreamer-edge-common.h (hunk 2 above)
Comment (jaeyun-jung) [reply src/libnnstreamer-edge/nnstreamer-edge-common.h:68]: Suggestion: adding lock in edge-data handle.
- metadata is internal struct.
- metadata is internally used in edge handle and edge-data handle, then called with edge handle lock. However edge-data handle does not have lock now.

File: src/libnnstreamer-edge/nnstreamer-edge-common.c [hunk 3]
```diff
@@ -138,6 +138,276 @@ nns_edge_strdup_printf (const char *format, ...)
...
+    SAFE_FREE (node->value);
+    node->value = val;
+    return NNS_EDGE_ERROR_NONE;
+  }
+
+  node = malloc (sizeof (nns_edge_metadata_node_s));
+  if (!node)
+    return NNS_EDGE_ERROR_OUT_OF_MEMORY;
+
+  memset (node, 0, sizeof (nns_edge_metadata_node_s));
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply src/libnnstreamer-edge/nnstreamer-edge-common.c:242]: updated

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c [hunk 4]
```diff
@@ -323,25 +338,68 @@ _nns_edge_cmd_receive (nns_edge_conn_s * conn, nns_edge_cmd_s * cmd)
...
+    if (!_receive_raw_data (conn->socket, cmd->meta, cmd->info.meta_size)) {
+      nns_edge_loge ("Failed to receive metadata from socket.");
+      ret = NNS_EDGE_ERROR_IO;
+      goto error;
+    }
+  }
+
+error:
   if (ret != NNS_EDGE_ERROR_NONE) {
     for (i = 0; i < n; i++) {
```
[the remark this answers is not in the document]
Comment (jaeyun-jung) [reply src/libnnstreamer-edge/nnstreamer-edge-internal.c:368]: Updated

File: src/libnnstreamer-edge/nnstreamer-edge-common.c (hunk 1 above)
Comment (jaeyun-jung) [reply src/libnnstreamer-edge/nnstreamer-edge-common.c:301]: updated

File: src/libnnstreamer-edge/nnstreamer-edge-common.h (hunk 2 above)
Comment (myungjoo) [reply src/libnnstreamer-edge/nnstreamer-edge-common.h:68]: That's be ok.

**Review by myungjoo — APPROVED** [review]
LGTM. PTAL: @again4you

**Comment by myungjoo** [comment]
Anyway, please hide internal data structure from other .C files. The internal data structure of a handle should not be exposed to its users.

File: src/libnnstreamer-edge/nnstreamer-edge-common.h (hunk 2 above)
Comment (again4you) [reply src/libnnstreamer-edge/nnstreamer-edge-common.h:68]: I agreed with jaeyun-jung  's opinion. 👍

**Pull request merged**